lib/spack/spack/util/git.py

```diff
@@ -164,3 +187,163 @@ def get_commit_sha(path: str, ref: str) -> Optional[str]:
             continue
 
     return None
+
+
+class VersionConditionalOption:
+    def __init__(self, key, value=None, min_version=(0, 0, 0), max_version=(99, 99, 99)):
+        self.key = key
+        self.value = value
+        self.min_version = min_version
+        self.max_version = max_version
+
+    def __call__(self, exe_version, value=None) -> List:
+        if (self.min_version <= exe_version) and (self.max_version >= exe_version):
+            option = [self.key]
+            if value:
+                option.append(value)
+            elif self.value:
+                option.append(self.value)
+            return option
+        else:
+            return []
+
+
+# The earliest git version where we start trying to optimizie clones
+# git@1.8.5 is when branch could also accept tag so we don't have to track ref types as closely
+MIN_OPT_VERSION = (1, 8, 5, 2)
+
+
+# Technically the flags existed earlier but we are pruning our logic to 1.8.5 or greater
+BRANCH = VersionConditionalOption("--branch", min_version=MIN_OPT_VERSION)
+SINGLE_BRANCH = VersionConditionalOption("--single-branch", min_version=MIN_OPT_VERSION)
+NO_SINGLE_BRANCH = VersionConditionalOption("--no-single-branch", min_version=MIN_OPT_VERSION)
+# Depth was introduced in 1.7.11 but isn't worth much without the --branch options
+DEPTH = VersionConditionalOption("--depth", 1, min_version=MIN_OPT_VERSION)
+
+FILTER_BLOB_NONE = VersionConditionalOption("--filter=blob:none", min_version=(2, 19, 0))
+NO_CHECKOUT = VersionConditionalOption("--no-checkout", min_version=(2, 34, 0))
+# technically sparse-checkout was added in 2.25, but we go forward since the model we use only
+# works with the `--cone` option
+SPARSE_CHECKOUT = VersionConditionalOption("sparse-checkout", "set", min_version=(2, 34, 0))
+
+
+def _exec_git_commands(git_exe, cmds, debug, dest=None):
+    if dest:
+        dest_cmd = ["-C", dest]
+        cmds = [dest_cmd + cmd for cmd in cmds]
+    for cmd in cmds:
+        if not debug:
+            # swallow extra output that leasks to str for non-debug case
+            git_exe(*cmd, error=str)
+        else:
+            git_exe(*cmd)
+
+
+def _exec_git_commands_unique_dir(git_exe, cmds, debug, dest=None):
+    if dest:
+        # mimic creating a dir and clean up if there is a failure like git clone
+        assert not os.path.isdir(dest)
+        os.mkdir(dest)
+        try:
+            _exec_git_commands(git_exe, cmds, debug, dest)
+        except exe.ProcessError:
+            shutil.rmtree(
+                dest, ignore_errors=False, onerror=fs.readonly_file_handler(ignore_errors=True)
+            )
+            raise
+    else:
+        _exec_git_commands(git_exe, cmds, debug, dest)
```

Comment:
If this is for mimicking `clone` we probably don't need to support the `dest=None` case for doing it in the current directory, since that's not a case that mimics any clone behavior.

At that point, this might become small enough to inline in the one method that calls it, but not certain on that.